### PR TITLE
Turn `lite::types::Requester` into an async trait

### PR DESCRIPTION
**Pull request by romac**
Closes: #170

On top of the changes mentioned in the linked issue, `lite::verifier::verify_bisection_inner` had to be manually adapted to return a boxed future because recursive async fn are not currently supported in Rust.

An possible alternative would be to rewrite `verify_bisection_inner` to not be recursive anymore, but that is outside of the scope of this PR.

* [x] Referenced an issue explaining the need for the change
* [ ] Updated all relevant documentation in docs
* [ ] Updated all code comments where relevant
* [ ] Wrote tests
* [ ] Updated CHANGES.md

### CODE REVIEW

File: tendermint/src/lite/verifier.rs [hunk 1]
```diff
@@ -333,75 +337,81 @@ where
...
+fn verify_bisection_inner<'a, H, C, L, R>(
+    trusted_state: &'a TrustedState<C, H>,
     untrusted_height: Height,
     trust_threshold: L,
-    req: &R,
-    mut cache: &mut Vec<TrustedState<C, H>>,
-) -> Result<TrustedState<C, H>, Error>
+    req: &'a R,
+    mut cache: &'a mut Vec<TrustedState<C, H>>,
+) -> LocalBoxFuture<'a, Result<TrustedState<C, H>, Error>>
```
Comment (Liamsi) [inline tendermint/src/lite/verifier.rs]: It is a bit unfortunate that the we had to change the method signature to return a boxed future. But this change is good in the sense that it surfaces that there is still a too tight coupling between verification logic and the requester (or networking / IO).

@brapse also suggested yesterday that we should probably implement the iterative version here too. It seems like this is a good idea for several reasons now.

File: tendermint-lite/src/main.rs [hunk 2]
```diff
@@ -7,35 +7,25 @@ use tendermint::{block::Height, Hash};
...
-        .basic_scheduler()
-        .enable_all()
-        .build()
-        .unwrap()
-        .block_on(future)
-}
-
-fn main() {
+#[tokio::main]
+async fn main() {
```
Comment (Liamsi) [inline tendermint-lite/src/main.rs]: I can look into incorporating these changes into the abscissa version of this main (in #125) likely with `abscissa_tokio` as @tarcieri suggested.

File: tendermint-lite/Cargo.toml [hunk 3]
```diff
@@ -8,4 +8,6 @@ edition = "2018"
 
 [dependencies]
 tendermint = { path = "../tendermint" }
-tokio = "0.2"
+
+async-trait = "0.1"
+tokio = { version = "0.2", features = ["full"] }
```
Comment (Liamsi) [inline tendermint-lite/Cargo.toml]: Just to double check: do we need _all_ Tokio public API features? Is there a good reason to enable them all anyways?

File: tendermint/src/lite/verifier.rs (hunk 1 above)
Comment (Liamsi) [reply tendermint/src/lite/verifier.rs]: https://github.com/interchainio/tendermint-rs/issues/174

File: tendermint-lite/Cargo.toml (hunk 3 above)
Comment (romac) [reply tendermint-lite/Cargo.toml]: Yeah sorry, `macros` should be sufficient (as in #173).

File: tendermint/src/lite/verifier.rs [hunk 4]
```diff
@@ -566,8 +579,8 @@ mod tests {
     }
 
     // valid to skip, but invalid commit. 1 validator.
-    #[test]
-    fn test_verify_single_skip_1_val_verify() {
+    #[tokio::test]
+    async fn test_verify_single_skip_1_val_verify() {
```
Comment (ebuchman) [inline tendermint/src/lite/verifier.rs]: Do the tests which don't use async functions need to be async?

Comment (Liamsi) [reply tendermint/src/lite/verifier.rs]: No, they don't. Maybe this wasn't intentional (only the bisection test use async functions).

Comment (Liamsi) [reply tendermint/src/lite/verifier.rs]: https://github.com/informalsystems/tendermint-rs/pull/195

**Pull request merged**